// next-page-tester/dist/getNextFiles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadErrorPageFiles = exports.loadExistingPageFiles = void 0;
const _document_1 = require("./_document");
const _app_1 = require("./_app");
const loadFile_1 = require("./loadFile");
const utils_1 = require("./utils");
const server_1 = require("./server");
// Get Document, App and Page files
function loadPageFiles({ absolutePagePath, options, }) {
    const { wrappers } = options;
    return {
        documentFile: (0, _document_1.getDocumentFile)({ options }),
        appFile: (0, _app_1.getAppFile)({ options }),
        pageFile: (0, loadFile_1.loadFile)({
            absolutePath: absolutePagePath,
        }),
        wrappersFile: wrappers
            ? (0, loadFile_1.loadFile)({
                absolutePath: wrappers,
            })
            : undefined,
    };
}
function loadExistingPageFiles({ absolutePagePath, options, }) {
    return {
        client: loadPageFiles({ absolutePagePath, options }),
        server: (0, server_1.executeAsIfOnServerSync)(() => (0, utils_1.executeWithFreshModules)(() => loadPageFiles({ absolutePagePath, options }), options)),
    };
}
exports.loadExistingPageFiles = loadExistingPageFiles;
function loadErrorPageFiles({ absolutePagePath, options, }) {
    return {
        client: loadPageFiles({ absolutePagePath, options }),
        server: (0, server_1.executeAsIfOnServerSync)(() => (0, utils_1.executeWithFreshModules)(() => loadPageFiles({ absolutePagePath, options }), options)),
    };
}
exports.loadErrorPageFiles = loadErrorPageFiles;
